packages/zapp/console/src/components/Launch/LaunchForm/MapInput.tsx

```diff
@@ -0,0 +1,110 @@
+import { FormControl, FormHelperText, TextField } from '@material-ui/core';
+import { makeStyles, Theme } from '@material-ui/core/styles';
+import * as React from 'react';
+import { requiredInputSuffix } from './constants';
+import { InputProps, InputType } from './types';
+import { formatType, getLaunchInputId, parseMappedTypeValue } from './utils';
+
+const useStyles = makeStyles((theme: Theme) => ({
+  formControl: {
+    width: '100%',
+  },
+  controls: {
+    width: '100%',
+    display: 'flex',
+    alignItems: 'center',
+    flexDirection: 'row',
+  },
+  keyControl: {
+    marginRight: theme.spacing(1),
+  },
+  valueControl: {
+    flexGrow: 1,
+  },
+}));
+
+/** Handles rendering of the input component for any primitive-type input */
+export const MapInput: React.FC<InputProps> = (props) => {
+  const {
+    error,
+    name,
+    onChange,
+    value = '',
+    typeDefinition: { subtype },
+  } = props;
+  const hasError = !!error;
+  const helperText = hasError ? error : props.helperText;
+  const classes = useStyles();
+
+  const { key: mapKey, value: mapValue } = parseMappedTypeValue(value);
+
+  const valueError = error?.startsWith("Value's value");
+
+  const handleKeyChange = React.useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      onChange(JSON.stringify({ [e.target.value || '']: mapValue }));
+    },
+    [mapValue],
+  );
+
+  const handleValueChange = React.useCallback(
+    (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+      onChange(JSON.stringify({ [mapKey]: e.target.value || '' }));
+    },
+    [mapKey],
+  );
+
+  const keyControl = (
+    <TextField
+      error={hasError && !valueError}
+      id={`${getLaunchInputId(name)}-key`}
+      label={`string${requiredInputSuffix}`}
+      onChange={handleKeyChange}
+      value={mapKey}
+      variant="outlined"
+      className={classes.keyControl}
+    />
+  );
+  let valueControl: JSX.Element;
+
+  switch (subtype?.type) {
+    case InputType.String:
+    case InputType.Integer:
+      valueControl = (
+        <TextField
+          error={valueError}
+          id={`${getLaunchInputId(name)}-value`}
```

Comment:
I see you are adding ids in ost of your components. Just curious what is the reasoning? Is it for unit test find by id?

Reply:
just followed the other parts of the code.. no specific reason